File: src/main/scala/pl/msulima/vistula/transpiler/rpn/expression/reference/Assign.scala
package pl.msulima.vistula.transpiler.rpn.expression.reference

import pl.msulima.vistula.parser.Ast
import pl.msulima.vistula.transpiler.rpn.{Operation, _}

object Assign extends Operator {

  def apply: PartialFunction[Ast.stmt, Token] = {
    case Ast.stmt.AssignStmt(expr, value) =>
      Operation(Assign, Seq(Tokenizer.boxed(value)), Tokenizer.apply(expr))
  }

  override def apply(operands: List[Constant], output: Constant): Constant = {
    Constant(s"${operands.head.value}.rxForEachOnce($$arg => ${output.value}.rxPush($$arg))")
  }
}

File: src/test/resources/pl/msulima/vistula/transpiler/reference/assign.js
X.rxMap($arg => ($arg + 2)).rxForEachOnce($arg => A.rxPush($arg));
vistula.constantObservable(false).rxForEachOnce($arg => B.rxPush($arg));
vistula.constantObservable(2 + 3).rxForEachOnce($arg => C.rxFlatMap($arg => ($arg.D)).rxPush($arg));
